### ISSUE TITLE
Support for `x instanceof Object`

### ISSUE THREAD

**Issue by philip-peterson** [facebook/flow#2132]
(This is probably related to #1690)

To check whether a certain value `x` is an object, I would like to use `x instanceof Object` because `typeof x === 'object'` ...
1. does not cover the case where x is `null`
2. does not cover functions being objects

**Comment by samwgoldman** [comment]
Can you share an example of some code that doesn't currently work? I'm not really sure what this issue is asking for.

**Comment by philip-peterson** [comment]
OK, so here's my use-case. I want to write the following function:

```
/* @flow */

type FooColor =
    string |
    {
        color1?: ?string,
    };

type Foo = {
    color: FooColor,
};

const s: Foo = {color: {color1: '#fff'}};

const extractColor = (generic: Foo): string => {
    return (generic.color instanceof Object ? generic.color.color1 : generic.color) || '';
};

console.log(extractColor(s));
```

but this fails Flow with the following error:
```
 16:     return (generic.color instanceof Object ? generic.color.color1 : generic.color) || '';
                                                                          ^^^^^^^^^^^^^ object type. This type is incompatible with the expected return type of
 15: const extractColor = (generic: Foo): string => {
                                          ^^^^^^ string
```

The fix is to change the `instanceof` check to a `typeof` check:

```
    [snipped]
    return (typeof generic.color === 'object' ? generic.color.color1 : generic.color) || '';
```

Which works with the given example, but `x instanceof Object` is more permissive than `typeof x === 'object'` in a way that I want, e.g. it allows x to be a function, or a symbol, or any future `typeof` result representing a JS value that can have a `.color1` property. 

Also `instanceof` conveniently returns false for null (`typeof` returns `'object'` for null, and for some strange reason Flow didn't force a null check, but that's probably a whole other issue).

Basically it boils down to me not wanting to write `typeof x === 'object' || typeof x === 'symbol' || typeof x === 'function'` any time I want to check whether a value can be dereferenced. (And not wanting to maintain that conditional everywhere.)

**Comment by philip-peterson** [comment]
I actually realized an alternative reason this might be happening, will explain in the morning...

**Comment by philip-peterson** [comment]
So I think what must be happening here is that when I use `instanceof Object`, that doesn't actually narrow down the possible types of `generic.color` like I was intending -- because a string is also an `instanceof Object`. And also the `=== null` check was most likely because `generic` was explicitly typed as a `Foo` rather than a `?Foo`.

Sorry about the confusion, this is a non-issue — and have a good day!

**Comment by samwgoldman** [comment]
Actually, `"hello" instanceof Object` is false. I think this is a bug in the way we handle this particular refinement. Thanks for raising the issue.

**Comment by samwgoldman** [comment]
Furthermore, it seems to only be broken in the negative/alternative branch:

```js
/* @flow */

type T = string | Object;

function f(x: T): string {
  if (x instanceof Object) {
    return "";
  } else {
    return x; // unexpected error: Object ~> string
  }
}

function g(x: T): Object {
  if (x instanceof Object) {
    return x;
  } else {
    return {};
  }
}
```

**Comment by avikchaudhuri** [comment]
@samwgoldman I bet this is because `Object` is treated specially, so in `string | Object` it is not the same internal type as in `x instanceof Object`.
